File: packages/android/captcha-sdk/src/main/java/com/captcha/pro/core/CaptchaGenerator.kt
package com.captcha.pro.core

import android.graphics.*
import android.util.Base64
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/**
 * Captcha generator for Android - Backend only
 */
class CaptchaGenerator {

    private var currentCaptchaId: String? = null
    private var currentTimestamp: Long = 0

    /**
     * Generate captcha from backend
     */
    suspend fun generate(options: CaptchaOptions): CaptchaResult {
        val response = resolveGetCaptcha(options)

        val data = response.data
        currentCaptchaId = data.captchaId
        currentTimestamp = data.timestamp ?: System.currentTimeMillis()

        val bgBitmap = withContext(Dispatchers.IO) {
            loadImageFromUrlOrBase64(data.bgImage)
        }

        val sliderBitmap = data.sliderImage?.let { sliderUrl ->
            withContext(Dispatchers.IO) {
                loadImageFromUrlOrBase64(sliderUrl)
            }
        }

        val targetPoints = mutableListOf<CaptchaPoint>()
        data.sliderY?.let { y ->
            targetPoints.add(CaptchaPoint(x = 0f, y = y, text = null))
        }

        return CaptchaResult(
            bgBitmap = bgBitmap,
            sliderBitmap = sliderBitmap,
            targetPoints = targetPoints,
            sliderY = data.sliderY ?: 0f,
            clickTexts = data.clickTexts,
            clickCharImages = data.clickCharImages,
            captchaId = data.captchaId,
            timestamp = currentTimestamp
        )
    }

    /**
     * Resolve getCaptcha endpoint: URL -> HTTP GET, Fn -> invoke.
     */
    private suspend fun resolveGetCaptcha(options: CaptchaOptions): BackendCaptchaResponse =
        withContext(Dispatchers.IO) {
            when (val ep = options.backendVerify.getCaptcha) {
                is GetCaptchaEndpoint.Url -> parseCaptchaResponse(
                    httpGetJson(buildCaptchaUrl(ep.url, options), options.backendVerify.headers, options.backendVerify.timeout)
                )
                is GetCaptchaEndpoint.Fn -> ep.fn()
            }
        }

    /**
     * Build getCaptcha URL with query params (type/width/height + slider|click dims).
     * Mirrors taro-vue BackendCaptchaParams query.
     */
    private fun buildCaptchaUrl(base: String, options: CaptchaOptions): String {
        val params = linkedMapOf(
            "type" to options.type.name.lowercase(),
            "width" to options.width.toString(),
            "height" to options.height.toString()
        )
        when (options.type) {
            CaptchaType.SLIDER -> {
                params["sliderWidth"] = options.sliderWidth.toString()
                params["sliderHeight"] = options.sliderHeight.toString()
            }
            CaptchaType.CLICK -> {
                params["clickCount"] = options.clickCount.toString()
            }
        }
        val query = params.entries.joinToString("&") { (k, v) -> "$k=$v" }
        val sep = if (base.contains("?")) "&" else "?"
        return "$base$sep$query"
    }

    /**
     * Resolve verify endpoint: URL -> HTTP POST, Fn -> invoke.
     */
    private suspend fun resolveVerify(data: CaptchaData, options: CaptchaOptions): BackendVerifyResponse =
        withContext(Dispatchers.IO) {
            when (val ep = options.backendVerify.verify) {
                is VerifyEndpoint.Url -> parseVerifyResponse(
                    httpPostJson(ep.url, serializeCaptchaData(data), options.backendVerify.headers, options.backendVerify.timeout)
                )
                is VerifyEndpoint.Fn -> ep.fn(data)
            }
        }

    private fun httpGetJson(url: String, headers: Map<String, String>?, timeout: Long): String {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = timeout.toInt()
            readTimeout = timeout.toInt()
            requestMethod = "GET"
            headers?.forEach { (k, v) -> setRequestProperty(k, v) }
            setRequestProperty("Accept", "application/json")
        }
        try {
            return conn.inputStream.bufferedReader().use { it.readText() }
        } finally {
            conn.disconnect()
        }
    }

    private fun httpPostJson(url: String, body: String, headers: Map<String, String>?, timeout: Long): String {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = timeout.toInt()
            readTimeout = timeout.toInt()
            requestMethod = "POST"
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json")
            headers?.forEach { (k, v) -> setRequestProperty(k, v) }
        }
        try {
            conn.outputStream.bufferedWriter().use { it.write(body) }
            return conn.inputStream.bufferedReader().use { it.readText() }
        } finally {
            conn.disconnect()
        }
    }

    private fun parseCaptchaResponse(json: String): BackendCaptchaResponse {
        val obj = JSONObject(json)
        val success = obj.optBoolean("success", true)
        if (!success) throw RuntimeException(obj.optString("message", "Failed to get captcha"))
        val dataObj = obj.optJSONObject("data") ?: obj
        val data = CaptchaResponseData(
            captchaId = dataObj.getString("captchaId"),
            bgImage = dataObj.getString("bgImage"),
            sliderImage = dataObj.optString("sliderImage").takeIf { it.isNotEmpty() },
            sliderY = dataObj.optDouble("sliderY").takeIf { !it.isNaN() }?.toFloat(),
            clickTexts = dataObj.optJSONArray("clickTexts")?.let { arr -> List(arr.length()) { arr.getString(it) } },
            clickCharImages = dataObj.optJSONArray("clickCharImages")?.let { arr -> List(arr.length()) { arr.getString(it) } },
            width = dataObj.optInt("width").takeIf { it != 0 },
            height = dataObj.optInt("height").takeIf { it != 0 },
            timestamp = dataObj.optDouble("timestamp").takeIf { it > 0 }?.toLong()
        )
        return BackendCaptchaResponse(data)
    }

    private fun parseVerifyResponse(json: String): BackendVerifyResponse {
        val obj = JSONObject(json)
        return BackendVerifyResponse(
            success = obj.optBoolean("success", false),
            message = obj.optString("message").takeIf { it.isNotEmpty() },
            data = obj.optJSONObject("data")?.let { d -> d.keys().asSequence().associateWith { d.get(it) } }
        )
    }

    private fun serializeCaptchaData(data: CaptchaData): String {
        val obj = JSONObject()
        obj.put("captchaId", data.captchaId)
        obj.put("type", when (data.type) {
            CaptchaType.SLIDER -> "slider"
            CaptchaType.CLICK -> "click"
        })
        val targetArr = JSONArray()
        data.target.forEach { item ->
            when (item) {
                is Float -> targetArr.put(item.toDouble())
                is Int -> targetArr.put(item)
                is Double -> targetArr.put(item)
                is CaptchaPoint -> {
                    val p = JSONObject()
                    p.put("x", item.x); p.put("y", item.y)
                    item.text?.let { p.put("text", it) }
                    targetArr.put(p)
                }
                else -> targetArr.put(item.toString())
            }
        }
        obj.put("target", targetArr)
        return obj.toString()
    }

    /**
     * Load image from URL or base64 string (public for views rendering char images)
     */
    fun loadImage(source: String): Bitmap = loadImageFromUrlOrBase64(source)

    /**
     * Load image from URL or base64 string
     */
    private fun loadImageFromUrlOrBase64(source: String): Bitmap {
        return when {
            source.startsWith("http://") || source.startsWith("https://") -> {
                URL(source).openStream().use { BitmapFactory.decodeStream(it) }
            }
            source.startsWith("data:image") -> {
                val base64Data = source.substringAfter(",")
                val bytes = Base64.decode(base64Data, Base64.DEFAULT)
                BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
            }
            else -> {
                try {
                    val bytes = Base64.decode(source, Base64.DEFAULT)
                    BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                } catch (e: Exception) {
                    throw IllegalArgumentException("Invalid image source: $source")
                }
            }
        }
    }

    /**
     * Get captcha data for verification.
     * target is polymorphic: SLIDER -> [sliderX] (Float); CLICK -> targetPoints.
     * Mirrors taro-vue target: number[] | Point[]
     */
    fun getCaptchaData(type: CaptchaType, targetPoints: List<CaptchaPoint> = emptyList(), sliderX: Float = 0f): CaptchaData {
        val target: List<Any> = when (type) {
            CaptchaType.SLIDER -> listOf(sliderX)
            CaptchaType.CLICK -> targetPoints
        }
        return CaptchaData(
            type = type,
            captchaId = currentCaptchaId ?: "",
            target = target,
            timestamp = currentTimestamp
        )
    }

    /**
     * Backend verify - resolve endpoint (URL -> HTTP, Fn -> invoke)
     */
    suspend fun backendVerify(data: CaptchaData, options: CaptchaOptions): BackendVerifyResponse {
        return resolveVerify(data, options)
    }

    /**
     * Convert bitmap to base64 string
     */
    fun bitmapToBase64(bitmap: Bitmap): String {
        val output = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, output)
        val bytes = output.toByteArray()
        return "data:image/png;base64," + Base64.encodeToString(bytes, Base64.NO_WRAP)
    }
}

File: packages/ios/Sources/Core/CaptchaGenerator.swift
import Foundation
import CoreGraphics
import UIKit

/// Captcha generator for iOS - Backend only
public class CaptchaGenerator {

    private var currentCaptchaId: String?
    private var currentTimestamp: TimeInterval = 0

    /// Generate captcha from backend
    public func generate(options: CaptchaOptions) async throws -> CaptchaResult {
        let response = try await resolveGetCaptcha(options: options)

        let data = response.data
        currentCaptchaId = data.captchaId
        currentTimestamp = data.timestamp ?? Date().timeIntervalSince1970 * 1000

        let bgImage = try await loadImage(from: data.bgImage)
        let sliderImage = try await data.sliderImage.map { try await loadImage(from: $0) }

        var targetPoints: [CaptchaPoint] = []
        if let sliderY = data.sliderY {
            targetPoints.append(CaptchaPoint(x: 0, y: sliderY))
        }

        return CaptchaResult(
            bgImage: bgImage,
            sliderImage: sliderImage,
            targetPoints: targetPoints,
            sliderY: data.sliderY ?? 0,
            clickTexts: data.clickTexts,
            clickCharImages: data.clickCharImages,
            captchaId: data.captchaId,
            timestamp: currentTimestamp
        )
    }

    /// Resolve getCaptcha endpoint: URL -> HTTP GET, function -> invoke.
    private func resolveGetCaptcha(options: CaptchaOptions) async throws -> BackendCaptchaResponse {
        switch options.backendVerify.getCaptcha {
        case .url(let url):
            let data = try await httpRequest(url: url, method: "GET", body: nil, headers: options.backendVerify.headers, timeout: options.backendVerify.timeout)
            return try parseCaptchaResponse(json: data)
        case .function(let fn):
            return try await fn()
        }
    }

    /// Resolve verify endpoint: URL -> HTTP POST, function -> invoke.
    private func resolveVerify(data: CaptchaData, options: CaptchaOptions) async throws -> BackendVerifyResponse {
        switch options.backendVerify.verify {
        case .url(let url):
            let body = try serializeCaptchaData(data: data)
            let json = try await httpRequest(url: url, method: "POST", body: body, headers: options.backendVerify.headers, timeout: options.backendVerify.timeout)
            return try parseVerifyResponse(json: json)
        case .function(let fn):
            return try await fn(data)
        }
    }

    private func httpRequest(url: String, method: String, body: Data?, headers: [String: String]?, timeout: TimeInterval) async throws -> Data {
        guard let url = URL(string: url) else { throw CaptchaError.invalidURL }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = timeout
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if method == "POST" { req.setValue("application/json", forHTTPHeaderField: "Content-Type") }
        req.httpBody = body
        headers?.forEach { req.setValue($1, forHTTPHeaderField: $0) }
        let (data, _) = try await URLSession.shared.data(for: req)
        return data
    }

    private func parseCaptchaResponse(json: Data) throws -> BackendCaptchaResponse {
        guard let obj = try JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            throw CaptchaError.networkError
        }
        if let success = obj["success"] as? Bool, !success {
            throw CaptchaError.networkError
        }
        let d = obj["data"] as? [String: Any] ?? obj
        let data = CaptchaResponseData(
            captchaId: d["captchaId"] as? String ?? "",
            bgImage: d["bgImage"] as? String ?? "",
            sliderImage: d["sliderImage"] as? String,
            sliderY: (d["sliderY"] as? Double).map { CGFloat($0) },
            clickTexts: d["clickTexts"] as? [String],
            clickCharImages: d["clickCharImages"] as? [String],
            width: d["width"] as? Int,
            height: d["height"] as? Int,
            timestamp: d["timestamp"] as? Double
        )
        return BackendCaptchaResponse(data: data)
    }

    private func parseVerifyResponse(json: Data) throws -> BackendVerifyResponse {
        guard let obj = try JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            throw CaptchaError.networkError
        }
        return BackendVerifyResponse(
            success: obj["success"] as? Bool ?? false,
            message: obj["message"] as? String,
            data: obj["data"] as? [String: Any]
        )
    }

    private func serializeCaptchaData(data: CaptchaData) throws -> Data {
        var target: [Any] = []
        for item in data.target {
            if let p = item as? CaptchaPoint {
                var dict: [String: Any] = ["x": p.x, "y": p.y]
                if let t = p.text { dict["text"] = t }
                target.append(dict)
            } else if let n = item as? CGFloat {
                target.append(n)
            } else if let n = item as? Double {
                target.append(n)
            }
        }
        let body: [String: Any] = [
            "captchaId": data.captchaId,
            "type": data.type.rawValue,
            "target": target,
            "timestamp": data.timestamp,
        ]
        return try JSONSerialization.data(withJSONObject: body)
    }

    /// Load image from URL or base64 string (public for views rendering char images)
    public func loadImage(from source: String) async throws -> UIImage {
        if source.starts(with: "http://") || source.starts(with: "https://") {
            guard let url = URL(string: source) else {
                throw CaptchaError.invalidURL
            }
            let (data, _) = try await URLSession.shared.data(from: url)
            guard let image = UIImage(data: data) else {
                throw CaptchaError.invalidImage
            }
            return image
        } else if source.starts(with: "data:image") {
            let base64Data = source.split(separator: ",").last?.data(using: .utf8) ?? Data()
            guard let data = Data(base64Encoded: base64Data),
                  let image = UIImage(data: data) else {
                throw CaptchaError.invalidImage
            }
            return image
        } else {
            // Try base64 directly
            guard let data = Data(base64Encoded: source),
                  let image = UIImage(data: data) else {
                throw CaptchaError.invalidImage
            }
            return image
        }
    }

    /// Get captcha data for verification.
    /// target is polymorphic: slider -> [sliderX] (CGFloat); click -> targetPoints.
    /// Mirrors taro-vue target: number[] | Point[]
    public func getCaptchaData(type: CaptchaType, targetPoints: [CaptchaPoint] = [], sliderX: CGFloat = 0) -> CaptchaData {
        let target: [Any]
        switch type {
        case .slider: target = [sliderX]
        case .click: target = targetPoints
        }
        return CaptchaData(
            type: type,
            captchaId: currentCaptchaId ?? "",
            target: target,
            timestamp: currentTimestamp
        )
    }

    /// Backend verify - resolve endpoint (URL -> HTTP, function -> invoke)
    public func backendVerify(data: CaptchaData, options: CaptchaOptions) async throws -> BackendVerifyResponse {
        return try await resolveVerify(data: data, options: options)
    }
}

public enum CaptchaError: Error {
    case invalidURL
    case invalidImage
    case networkError
    case verificationFailed
}
